### PR TITLE
Initial Boards+Channels implementation

### PR DESCRIPTION
**Pull request by jespino**

### CODE REVIEW

**Comment by jespino** [comment]
I have to address 3 TODOs there, but it is ready to be reviewed, nothing mayor is going to change.

File: mattermost-plugin/webapp/tests/utils.tsx [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import React from 'react'
+import {IntlProvider} from 'react-intl'
+
+export const wrapIntl = (children?: React.ReactNode): JSX.Element => (
+    <IntlProvider locale='en'>
+        {children}
+    </IntlProvider>
+)
```
Comment (sbishel) [inline mattermost-plugin/webapp/tests/utils.tsx:8]: We already have [wrapIntl](https://github.com/mattermost/focalboard/blob/main/webapp/src/testUtils.tsx#L12)

File: webapp/src/components/searchDialog/searchDialog.tsx [hunk 2]
```diff
@@ -19,6 +19,43 @@ type Props = {
...
 }
 
+export const EmptySearch = () => (
+    <div className='noResults introScreen'>
+        <div className='iconWrapper'>
+            <Search/>
+        </div>
+        <h4 className='text-heading4'>
+            <FormattedMessage
+                id='FindBoFindBoardsDialog.IntroText'
```
Comment (sbishel) [inline webapp/src/components/searchDialog/searchDialog.tsx:29]:
```suggestion
                id='FindBoardsDialog.IntroText'
```

File: webapp/src/components/shareBoard/shareBoard.tsx [hunk 3]
```diff
@@ -92,8 +97,9 @@ function isLastAdmin(members: BoardMember[]) {
 export default function ShareBoardDialog(props: Props): JSX.Element {
     const [wasCopiedPublic, setWasCopiedPublic] = useState(false)
     const [wasCopiedInternal, setWasCopiedInternal] = useState(false)
+    const [showLinkChannelConfirmation, setShowLinkChannelConfirmation] = useState<Channel|null>(null)
     const [sharing, setSharing] = useState<ISharing|undefined>(undefined)
-    const [selectedUser, setSelectedUser] = useState<IUser|null>(null)
+    const [selectedUser, setSelectedUser] = useState<IUser|any|null>(null)
```
Comment (sbishel) [inline webapp/src/components/shareBoard/shareBoard.tsx:102]: Is this change necessary?

File: webapp/src/components/shareBoard/shareBoard.tsx [hunk 4]
```diff
@@ -135,6 +141,17 @@ export default function ShareBoardDialog(props: Props): JSX.Element {
         await loadData()
     }
 
+    const onLinkBoard = async (channel: Channel, confirmed?: boolean) => {
+        if (channel.type === 'O' && !confirmed) {
```
Comment (sbishel) [inline webapp/src/components/shareBoard/shareBoard.tsx:145]: We have a constant for this...https://github.com/mattermost/focalboard/blob/main/webapp/src/blocks/board.ts#L11

File: webapp/src/components/shareBoard/shareBoard.tsx [hunk 5]
```diff
@@ -264,18 +281,36 @@ export default function ShareBoardDialog(props: Props): JSX.Element {
...
         return(
             <div className='user-item'>
-                {Utils.isFocalboardPlugin() &&
-                    <img
-                        src={Utils.getProfilePicture(user.id)}
-                        className='user-item__img'
-                    />
-                }
+                {channel.type === 'P' && <PrivateIcon/>}
+                {channel.type === 'O' && <PublicIcon/>}
```
Comment (sbishel) [inline webapp/src/components/shareBoard/shareBoard.tsx:311]: Constants already exist for these. https://github.com/mattermost/focalboard/blob/main/webapp/src/blocks/board.ts#L11

File: mattermost-plugin/webapp/src/components/rhsChannelBoards.tsx [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    const channelBoards = boards.filter((b) => b.channelId === currentChannel.id)
+
+    if (channelBoards.length === 0) {
+        return (
+            <div className='focalboard-body'>
+                <div className='RHSChannelBoards empty'>
+                    <h2>
+                        <FormattedMessage
+                            id='rhs-boards.no-boards-linked-to-channel'
+                            defaultMessage='No Boards are linked to {channelName} yet'
```
Comment (sbishel) [inline mattermost-plugin/webapp/src/components/rhsChannelBoards.tsx:46]:
```suggestion
                            defaultMessage='No boards are linked to {channelName} yet'
```

File: mattermost-plugin/webapp/src/components/rhsChannelBoards.tsx [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                    </div>
+                    <div className='boards-screenshots'><img src={boardsScreenshots}/></div>
+                    <Button
+                        onClick={() => dispatch(setLinkToChannel(currentChannel.id))}
+                        emphasis='primary'
+                        size='medium'
+                    >
+                        <FormattedMessage
+                            id='rhs-boards.link-boards-to-channel'
+                            defaultMessage='Link Boards to {channelName}'
```
Comment (sbishel) [inline mattermost-plugin/webapp/src/components/rhsChannelBoards.tsx:64]:
```suggestion
                            defaultMessage='Link boards to {channelName}'
```

File: mattermost-plugin/webapp/src/components/rhsChannelBoards.tsx [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    return (
+        <div className='focalboard-body'>
+            <div className='RHSChannelBoards'>
+                <div className='rhs-boards-header'>
+                    <span className='linked-boards'>
+                        <FormattedMessage
+                            id='rhs-boards.linked-boards'
+                            defaultMessage='Linked Boards'
```
Comment (sbishel) [inline mattermost-plugin/webapp/src/components/rhsChannelBoards.tsx:80]:
```suggestion
                            defaultMessage='Linked boards'
```

File: server/api/api.go [hunk 9]
```diff
@@ -2170,6 +2172,171 @@ func (a *API) handleGetTeamUsers(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	auditRec := a.makeAuditRecord(r, "searchMyChannels", audit.Fail)
+	defer a.audit.LogRecord(audit.LevelRead, auditRec)
+	auditRec.AddMeta("teamID", teamID)
+
+	// retrieve boards list
+	channels, err := a.app.SearchUserChannels(teamID, userID, searchQuery)
+	if err != nil {
+		fmt.Println(err)
```
Comment (sbishel) [inline server/api/api.go:2231]: Do you want to log here or remove?

File: server/api/api.go [hunk 10]
```diff
@@ -2170,6 +2172,171 @@ func (a *API) handleGetTeamUsers(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	auditRec := a.makeAuditRecord(r, "getChannel", audit.Fail)
+	defer a.audit.LogRecord(audit.LevelRead, auditRec)
+	auditRec.AddMeta("teamID", teamID)
+	auditRec.AddMeta("channelID", teamID)
+
+	channel, err := a.app.GetChannel(teamID, channelID)
+	if err != nil {
+		fmt.Println(err)
```
Comment (sbishel) [inline server/api/api.go:2313]: Do you want to log here?

File: server/services/store/mattermostauthlayer/mattermostauthlayer.go [hunk 11]
```diff
@@ -625,6 +625,200 @@ func (s *MattermostAuthLayer) GetCloudLimits() (*mmModel.ProductLimits, error) {
...
+func (s *MattermostAuthLayer) SearchUserChannels(teamID, userID, query string) ([]*mmModel.Channel, error) {
+	channels, err := s.pluginAPI.GetChannelsForTeamForUser(teamID, userID, false)
+	if err != nil {
+		return nil, err
+	}
+
+	result := []*mmModel.Channel{}
+	count := 0
+	for _, channel := range channels {
+		if channel.Type != "D" && channel.Type != "G" && (strings.Contains(channel.Name, query) || strings.Contains(channel.DisplayName, query)) {
```
Comment (sbishel) [inline server/services/store/mattermostauthlayer/mattermostauthlayer.go:803]: Channel types we have constants for.

File: server/services/store/sqlstore/sqlstore.go [hunk 12]
```diff
@@ -33,6 +33,14 @@ type SQLStore struct {
...
 type MutexFactory func(name string) (*cluster.Mutex, error)
 
+type NotSupportedError struct {
+	msg string
+}
+
+func (pe NotSupportedError) Error() string {
+	return pe.msg
+}
+
```
Comment (sbishel) [inline server/services/store/sqlstore/sqlstore.go:43]: I believe these are already defined.

**Review by sbishel — CHANGES_REQUESTED** [review]
Overall, looks good. I've just got some nits. I saw @justinegeffen put through some changes for second words being capitalized, (ie. "Linked Boards" vs "Linked boards") so I pointed out those I found. It's up to her if they need to fixed.

File: webapp/src/components/shareBoard/shareBoard.tsx (hunk 4 above)
Comment (jespino) [reply webapp/src/components/shareBoard/shareBoard.tsx:145]: Actually that is not exactly that, that are "board type" not channel type, I going to create a couple of constants for that in the store/channels.ts because is there where the Channel struct is defined.

File: server/api/api.go (hunk 9 above)
Comment (jespino) [reply server/api/api.go:2231]: clearly remove

File: server/services/store/sqlstore/sqlstore.go (hunk 12 above)
Comment (jespino) [reply server/services/store/sqlstore/sqlstore.go:43]: Actually, it is defined but in the mattermostauthlayer, so I change it to expose that error as part of the store package, and use it in the sqlstore and in the mattermostauthlayer.

**Comment by jespino** [comment]
@sbishel PTAL

**Comment by jespino** [comment]
Ignoring the cypress test failure and merging it.
